de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java

```diff
@@ -607,17 +648,39 @@ private synchronized void startNextProjectNegotiation() {
       return;
     }
 
+    List<User> recipients = new ArrayList<>();
+    if (session.isHost()) {
+      /*
+       * If we received these project from a non-host user previously,
+       * that user already has the project.
+       */
+      users:
+      for (User user : session.getRemoteUsers()) {
+        for (IProject project : projectSharingData) {
+          if (session.userHasProject(user, project)) {
```

Comment:
Oh god, partial sharing really is a feature that breaks so much. Any ideas how we could solve this one?

Before the Project was just shared with every remote user, so this would probably be fixed, if it would be shared with everyone, but the one that shared the Project with the host?

So in your scenario Alice would share it with Carl a second time, that should fix it, right?

Reply:
> So in your scenario Alice would share it with Carl a second time, that should fix it, right?

Yes, we have to start a negotiation like Bob did with Alice (including the new files)

I have a suggestion that would fix both scenarios: add `User originUser` to
`private void executeOutgoingProjectNegotiation(ProjectSharingData projectSharingData)`
replace the block in line 652:
``` java
List<User> recipients = new ArrayList<>();
if (session.isHost()) {
  for (User user : session.getRemoteUsers()) {
    if (user != originUser) recipients.add(user);
  }
} else {
  recipients.add(session.getHost());
}
```

and call like
https://gist.github.com/stefaus/b641b99753a035c7b357bdf5f2acb51c#file-sarossessionmanager-java-L156-L157
and
https://gist.github.com/stefaus/b641b99753a035c7b357bdf5f2acb51c#file-sarossessionmanager-java-L635

Reply:
That is what I had in mind as well. Added/Fixed